File: app/src-tauri/src/repo/subgoal.rs
use rusqlite::{params, Connection, Row};

use super::{now, optional_text, required_text, task};
use crate::error::{Error, Result};
use crate::models::{Subgoal, SubgoalDetail, SubgoalInput, SubgoalUpdate};
use crate::progress;

const COLUMNS: &str = "id, goal_id, title, due_date, is_complete, position, created_at, updated_at";
const ORDER: &str = "ORDER BY position, id";

fn map(row: &Row) -> rusqlite::Result<Subgoal> {
    Ok(Subgoal {
        id: row.get("id")?,
        goal_id: row.get("goal_id")?,
        title: row.get("title")?,
        due_date: row.get("due_date")?,
        is_complete: row.get("is_complete")?,
        position: row.get("position")?,
        created_at: row.get("created_at")?,
        updated_at: row.get("updated_at")?,
    })
}

/// Every subgoal in the database, grouped by goal. The Task Manager needs this
/// to offer subgoals as task parents without loading one goal at a time.
pub fn list_all(conn: &Connection) -> Result<Vec<Subgoal>> {
    let mut stmt = conn.prepare(&format!("SELECT {COLUMNS} FROM subgoals ORDER BY goal_id, position, id"))?;
    let subgoals = stmt.query_map([], map)?.collect::<rusqlite::Result<_>>()?;
    Ok(subgoals)
}

pub fn list_for_goal(conn: &Connection, goal_id: i64) -> Result<Vec<Subgoal>> {
    let mut stmt = conn.prepare(&format!(
        "SELECT {COLUMNS} FROM subgoals WHERE goal_id = ?1 {ORDER}"
    ))?;
    let subgoals = stmt
        .query_map(params![goal_id], map)?
        .collect::<rusqlite::Result<_>>()?;
    Ok(subgoals)
}

/// Each subgoal with its tasks and its own progress attached.
pub fn detail_for_goal(conn: &Connection, goal_id: i64) -> Result<Vec<SubgoalDetail>> {
    list_for_goal(conn, goal_id)?
        .into_iter()
        .map(|subgoal| {
            let tasks = task::list_for_subgoal(conn, subgoal.id)?;
            let completions: Vec<f64> = tasks.iter().map(|t| t.status.completion()).collect();
            Ok(SubgoalDetail {
                progress: progress::subgoal_progress(subgoal.is_complete, &completions),
                subgoal,
                tasks,
            })
        })
        .collect()
}

/// Just the progress numbers, for computing a parent goal's average.
pub fn progresses_for_goal(conn: &Connection, goal_id: i64) -> Result<Vec<f64>> {
    Ok(detail_for_goal(conn, goal_id)?
        .into_iter()
        .map(|detail| detail.progress)
        .collect())
}

pub fn get(conn: &Connection, id: i64) -> Result<Subgoal> {
    conn.query_row(
        &format!("SELECT {COLUMNS} FROM subgoals WHERE id = ?1"),
        params![id],
        map,
    )
    .map_err(|err| match err {
        rusqlite::Error::QueryReturnedNoRows => Error::not_found("subgoal", id),
        other => other.into(),
    })
}

pub fn create(conn: &Connection, input: SubgoalInput) -> Result<Subgoal> {
    let title = required_text("subgoal title", &input.title)?;
    super::goal::ensure_exists(conn, input.goal_id)?;

    let next_position: i64 = conn.query_row(
        "SELECT COALESCE(MAX(position) + 1, 0) FROM subgoals WHERE goal_id = ?1",
        params![input.goal_id],
        |row| row.get(0),
    )?;
    let timestamp = now();

    conn.execute(
        "INSERT INTO subgoals (goal_id, title, due_date, is_complete, position, created_at, updated_at)
         VALUES (?1, ?2, ?3, 0, ?4, ?5, ?5)",
        params![
            input.goal_id,
            title,
            optional_text(input.due_date),
            next_position,
            timestamp
        ],
    )?;

    get(conn, conn.last_insert_rowid())
}

pub fn update(conn: &Connection, id: i64, input: SubgoalUpdate) -> Result<Subgoal> {
    let title = required_text("subgoal title", &input.title)?;
    let changed = conn.execute(
        "UPDATE subgoals SET title = ?1, due_date = ?2, is_complete = ?3, updated_at = ?4
         WHERE id = ?5",
        params![
            title,
            optional_text(input.due_date),
            input.is_complete,
            now(),
            id
        ],
    )?;

    if changed == 0 {
        return Err(Error::not_found("subgoal", id));
    }
    get(conn, id)
}

pub fn set_complete(conn: &Connection, id: i64, is_complete: bool) -> Result<Subgoal> {
    let changed = conn.execute(
        "UPDATE subgoals SET is_complete = ?1, updated_at = ?2 WHERE id = ?3",
        params![is_complete, now(), id],
    )?;

    if changed == 0 {
        return Err(Error::not_found("subgoal", id));
    }
    get(conn, id)
}

/// Deleting a subgoal keeps its tasks; they fall back to being direct tasks of
/// the goal (ON DELETE SET NULL on `subgoal_id`).
pub fn delete(conn: &Connection, id: i64) -> Result<()> {
    let changed = conn.execute("DELETE FROM subgoals WHERE id = ?1", params![id])?;
    if changed == 0 {
        return Err(Error::not_found("subgoal", id));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db;
    use crate::models::{GoalInput, TaskInput, TaskStatus, Timeframe};
    use crate::repo::goal;

    fn seed_goal(conn: &Connection) -> i64 {
        goal::create(
            conn,
            GoalInput {
                category_id: None,
                title: "Ship v1".into(),
                description: None,
                timeframe: Timeframe::Mid,
                due_date: None,
                motivation_text: None,
                motivation_image_path: None,
            },
        )
        .unwrap()
        .id
    }

    fn add(conn: &Connection, goal_id: i64, title: &str) -> Subgoal {
        create(
            conn,
            SubgoalInput {
                goal_id,
                title: title.into(),
                due_date: None,
            },
        )
        .unwrap()
    }

    #[test]
    fn new_subgoals_append_in_order() {
        let conn = db::open_in_memory().unwrap();
        let goal_id = seed_goal(&conn);

        let first = add(&conn, goal_id, "Write the schema");
        let second = add(&conn, goal_id, "Wire the UI");

        assert_eq!((first.position, second.position), (0, 1));
        assert_eq!(list_for_goal(&conn, goal_id).unwrap().len(), 2);
    }

    #[test]
    fn list_all_spans_every_goal_in_position_order() {
        let conn = db::open_in_memory().unwrap();
        let first_goal = seed_goal(&conn);
        let second_goal = seed_goal(&conn);

        add(&conn, first_goal, "Write the schema");
        add(&conn, second_goal, "Book the flight");
        add(&conn, first_goal, "Wire the UI");

        let titles: Vec<String> = list_all(&conn)
            .unwrap()
            .into_iter()
            .map(|subgoal| subgoal.title)
            .collect();

        assert_eq!(titles, ["Write the schema", "Wire the UI", "Book the flight"]);
    }

    #[test]
    fn rejects_a_missing_goal() {
        let conn = db::open_in_memory().unwrap();
        let err = create(
            &conn,
            SubgoalInput {
                goal_id: 404,
                title: "Orphan".into(),
                due_date: None,
            },
        )
        .unwrap_err();

        assert_eq!(err.kind(), "not_found");
    }

    #[test]
    fn taskless_subgoal_progress_follows_its_checkbox() {
        let conn = db::open_in_memory().unwrap();
        let goal_id = seed_goal(&conn);
        let subgoal = add(&conn, goal_id, "Write the schema");

        assert_eq!(progresses_for_goal(&conn, goal_id).unwrap(), vec![0.0]);

        set_complete(&conn, subgoal.id, true).unwrap();
        assert_eq!(progresses_for_goal(&conn, goal_id).unwrap(), vec![1.0]);
    }

    #[test]
    fn subgoal_with_tasks_averages_them() {
        let conn = db::open_in_memory().unwrap();
        let goal_id = seed_goal(&conn);
        let subgoal = add(&conn, goal_id, "Write the schema");

        for title in ["Draft the tables", "Add indexes"] {
            task::create(
                &conn,
                TaskInput {
                    title: title.into(),
                    due_date: None,
                    goal_id: None,
                    subgoal_id: Some(subgoal.id),
                    is_recurring: false,
                },
            )
            .unwrap();
        }
        let first = task::list_for_subgoal(&conn, subgoal.id).unwrap()[0].id;
        task::set_status(&conn, first, TaskStatus::Done).unwrap();

        assert_eq!(progresses_for_goal(&conn, goal_id).unwrap(), vec![0.5]);
    }

    #[test]
    fn deleting_a_subgoal_leaves_its_tasks_on_the_goal() {
        let conn = db::open_in_memory().unwrap();
        let goal_id = seed_goal(&conn);
        let subgoal = add(&conn, goal_id, "Write the schema");
        task::create(
            &conn,
            TaskInput {
                title: "Draft the tables".into(),
                due_date: None,
                goal_id: None,
                subgoal_id: Some(subgoal.id),
                is_recurring: false,
            },
        )
        .unwrap();

        delete(&conn, subgoal.id).unwrap();

        let direct = task::list_direct_for_goal(&conn, goal_id).unwrap();
        assert_eq!(direct.len(), 1);
    }
}

File: app/src/lib/db/repo/subgoal.ts
import type { SqlDriver } from '../driver';
import { AppError } from '../error';
import type { Subgoal, SubgoalDetail, SubgoalInput, SubgoalUpdate } from '../../api/types';
import { now, optionalText, requiredText } from './helpers';
import * as task from './task';
import { countsTowardProgress, subgoalProgress, taskCompletion } from '../logic/progress';

const COLUMNS = 'id, goal_id, title, due_date, is_complete, position, created_at, updated_at';
const ORDER = 'ORDER BY position, id';

interface SubgoalRow {
	id: number;
	goal_id: number;
	title: string;
	due_date: string | null;
	is_complete: number;
	position: number;
	created_at: string;
	updated_at: string;
}

function map(row: SubgoalRow): Subgoal {
	return {
		id: row.id,
		goalId: row.goal_id,
		title: row.title,
		dueDate: row.due_date,
		isComplete: Boolean(row.is_complete),
		position: row.position,
		createdAt: row.created_at,
		updatedAt: row.updated_at
	};
}

/** Every subgoal in the database, grouped by goal. The Task Manager needs this
 * to offer subgoals as task parents without loading one goal at a time. */
export async function listAll(driver: SqlDriver): Promise<Subgoal[]> {
	const rows = await driver.select<SubgoalRow>(
		`SELECT ${COLUMNS} FROM subgoals ORDER BY goal_id, position, id`
	);
	return rows.map(map);
}

export async function listForGoal(driver: SqlDriver, goalId: number): Promise<Subgoal[]> {
	const rows = await driver.select<SubgoalRow>(
		`SELECT ${COLUMNS} FROM subgoals WHERE goal_id = ?1 ${ORDER}`,
		[goalId]
	);
	return rows.map(map);
}

/** Each subgoal with its tasks and its own progress attached. */
export async function detailForGoal(driver: SqlDriver, goalId: number): Promise<SubgoalDetail[]> {
	const subgoals = await listForGoal(driver, goalId);
	const details: SubgoalDetail[] = [];
	for (const subgoal of subgoals) {
		const tasks = await task.listForSubgoal(driver, subgoal.id);
		const completions = tasks.filter(countsTowardProgress).map((t) => taskCompletion(t.status));
		details.push({ ...subgoal, progress: subgoalProgress(subgoal.isComplete, completions), tasks });
	}
	return details;
}

/** Just the progress numbers, for computing a parent goal's average. */
export async function progressesForGoal(driver: SqlDriver, goalId: number): Promise<number[]> {
	return (await detailForGoal(driver, goalId)).map((detail) => detail.progress);
}

export async function get(driver: SqlDriver, id: number): Promise<Subgoal> {
	const rows = await driver.select<SubgoalRow>(`SELECT ${COLUMNS} FROM subgoals WHERE id = ?1`, [
		id
	]);
	const row = rows[0];
	if (!row) throw AppError.notFound('subgoal', id);
	return map(row);
}

export async function create(driver: SqlDriver, input: SubgoalInput): Promise<Subgoal> {
	const title = requiredText('subgoal title', input.title);
	const owner = await driver.select('SELECT id FROM goals WHERE id = ?1', [input.goalId]);
	if (owner.length === 0) throw AppError.notFound('goal', input.goalId);

	const nextPosition = await driver.select<{ next: number }>(
		'SELECT COALESCE(MAX(position) + 1, 0) as next FROM subgoals WHERE goal_id = ?1',
		[input.goalId]
	);
	const timestamp = now();

	const result = await driver.execute(
		`INSERT INTO subgoals (goal_id, title, due_date, is_complete, position, created_at, updated_at)
         VALUES (?1, ?2, ?3, 0, ?4, ?5, ?5)`,
		[input.goalId, title, optionalText(input.dueDate), nextPosition[0].next, timestamp]
	);

	return get(driver, result.lastInsertId);
}

export async function update(
	driver: SqlDriver,
	id: number,
	input: SubgoalUpdate
): Promise<Subgoal> {
	const title = requiredText('subgoal title', input.title);
	const result = await driver.execute(
		`UPDATE subgoals SET title = ?1, due_date = ?2, is_complete = ?3, updated_at = ?4
         WHERE id = ?5`,
		[title, optionalText(input.dueDate), Number(input.isComplete), now(), id]
	);
	if (result.rowsAffected === 0) throw AppError.notFound('subgoal', id);
	return get(driver, id);
}

export async function setComplete(
	driver: SqlDriver,
	id: number,
	isComplete: boolean
): Promise<Subgoal> {
	const result = await driver.execute(
		'UPDATE subgoals SET is_complete = ?1, updated_at = ?2 WHERE id = ?3',
		[Number(isComplete), now(), id]
	);
	if (result.rowsAffected === 0) throw AppError.notFound('subgoal', id);
	return get(driver, id);
}

/** Deleting a subgoal keeps its tasks; they fall back to being direct tasks of
 * the goal (ON DELETE SET NULL on subgoal_id). */
export async function remove(driver: SqlDriver, id: number): Promise<void> {
	const result = await driver.execute('DELETE FROM subgoals WHERE id = ?1', [id]);
	if (result.rowsAffected === 0) throw AppError.notFound('subgoal', id);
}
